network/p2p/unstaked_translator_test.go

```diff
@@ -0,0 +1,98 @@
+package p2p
+
+import (
+	"crypto/rand"
+	"math"
+	"testing"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/stretchr/testify/require"
+
+	fcrypto "github.com/onflow/flow-go/crypto"
+)
+
+// This test shows we can't use ECDSA P-256 keys for libp2p and expect PeerID <=> PublicKey bijections
+func TestIDTranslationP256(t *testing.T) {
+	loops := 50
+	for i := 0; i < loops; i++ {
+		pID := createPeerIDFromAlgo(t, fcrypto.ECDSAP256)
+
+		// check that we can not extract the public key back
+		// This makes sense: the x509 serialization of ECDSA P-256 keys in uncompressed form is 64 + 2 bytes,
+		// and libp2p uses multihash.IDENTITY only on serializations of less than 42 bytes
+		_, err := pID.ExtractPublicKey()
+		require.NotNil(t, err)
+
+	}
+}
+
+// This test shows we can use ECDSA Secp256k1 keys for libp2p and expect PeerID <=> PublicKey bijections
+func TestIDTranslationSecp256k1(t *testing.T) {
+	loops := 50
+	for i := 0; i < loops; i++ {
+		pID := createPeerIDFromAlgo(t, fcrypto.ECDSASecp256k1)
+
+		// check that we can extract the public key back
+		// This makes sense: the compressed serialization of ECDSA Secp256k1 keys is 33 + 2 bytes,
+		// and libp2p uses multihash.IDENTITY on serializations of less than 42 bytes
+		_, err := pID.ExtractPublicKey()
+		require.NoError(t, err)
+
+	}
+}
+
+func TestUnstakedTranslationRoundTrip(t *testing.T) {
+	loops := 50
+	unstakedTranslator := NewUnstakedNetworkIDTranslator()
+	for i := 0; i < loops; i++ {
+		pID := createPeerIDFromAlgo(t, fcrypto.ECDSASecp256k1)
+
+		pk, err := pID.ExtractPublicKey()
+		require.NoError(t, err)
+
+		// for a secp256k1 key, this is compressed representation preceded by 00 bits
+		// indicating the multihash.IDENTITY
+		pkBytes, err := pk.Raw()
+		require.NoError(t, err)
+
+		// key is positive, roundtrip should be possible
+		if pkBytes[0] == 0x02 {
```

Comment:
Can we enforce that this is actually run a certain number of times, instead of just running the loop 50 times in total and potentially (though hopefully not likely) never getting one with 0x02?

What I'm actually worried about is whether byte [0] is actually the correct byte to check here, or whether `pk.Raw()` actually returns some more bytes in the beginning.

For example, above you note

> for a secp256k1 key, this is compressed representation preceded by 00 bits indicating the multihash.IDENTITY

So would it be the case that `pkBytes[0] == 0x00` all the time? If that were true, this test would always pass because the if statement always fails.

This is something I was meaning to check and hence why I had it as a TODO above, but since the TODO is now being removed I'd like to make sure of it here.

Reply:
> Can we enforce that this is actually run a certain number of times, instead of just running the loop 50 times in total and potentially (though hopefully not likely) never getting one with 0x02?

Sure! Though the probability is 50%. I'll add the change.

> So would it be the case that pkBytes[0] == 0x00 all the time? 

I wrote bits, not byte.

Reply:
Thanks for checking :)